src/window/macos.rs

```diff
@@ -0,0 +1,152 @@
+    },
+    Foundation::{NSPoint, NSRect, NSSize},
+};
+use objc2::{declare_class, msg_send_id, mutability::InteriorMutable, rc::Id, ClassType};
+
+use csscolorparser::Color;
+use raw_window_handle::{HasRawWindowHandle, RawWindowHandle};
+use winit::window::Window;
+
+use crate::{cmd_line::CmdLineSettings, error_msg, frame::Frame, settings::SETTINGS};
+
+use super::WindowSettings;
+
+declare_class!(
+    // A view to simulate the double-click-to-zoom effect.
+    struct TitlebarClickHandler;
+
+    unsafe impl ClassType for TitlebarClickHandler {
+        type Super = NSView;
+        type Mutability = InteriorMutable;
+        const NAME: &'static str = "TitlebarClickHandler";
+    }
+
+    unsafe impl TitlebarClickHandler {
+        #[method(mouseDown:)]
+        unsafe fn mouse_down(&self, event: &NSEvent) {
+            if event.clickCount() == 2 {
+                self.window().unwrap().zoom(Some(&self));
+            }
+        }
+    }
+);
+
+impl TitlebarClickHandler {
+    pub fn new() -> Id<TitlebarClickHandler> {
+        unsafe { msg_send_id![Self::alloc(), init] }
+    }
+}
+
+lazy_static! {
+    static ref TITLEBAR_HEIGHT: f64 = MacosWindowFeature::calculate_titlebar_height();
+}
+
+// A trait for winit Window on macos, by which we can easily manipulate the window.
+pub struct MacosWindowFeature {
+    ns_window: Id<NSWindow>,
+    has_titlebar: bool,
+}
+
+impl MacosWindowFeature {
+    pub fn from_winit_window(window: &Window) -> MacosWindowFeature {
+        let ns_window = if let RawWindowHandle::AppKit(handle) = window.raw_window_handle() {
+            unsafe { Id::new(handle.ns_window as *mut NSWindow).unwrap() }
+        } else {
+            panic!("Not a appkit window.");
+        };
+
+        let frame = SETTINGS.get::<CmdLineSettings>().frame;
+        let has_titlebar = frame == Frame::Full || frame == Frame::Transparent;
+
+        // Frame::None => NSWindowStyleMaskBorderless
+        // Frame::Buttonless => Full Decoration + Hidden Titlebar to get a round corner window
+        // Frame::Full => Full Decoration
+        unsafe {
+            let full_frame_style_mask = NSWindowStyleMaskTitled
+                | NSWindowStyleMaskFullSizeContentView
+                | NSWindowStyleMaskClosable
+                | NSWindowStyleMaskMiniaturizable
+                | NSWindowStyleMaskResizable;
+
+            let style_mask = match frame {
+                Frame::None => NSWindowStyleMaskBorderless,
+                Frame::Buttonless => NSWindowStyleMaskTitled | NSWindowStyleMaskFullSizeContentView,
+                Frame::Transparent => full_frame_style_mask,
+                Frame::Full => full_frame_style_mask,
+            };
+
+            ns_window.setStyleMask(style_mask);
+            ns_window.setTitlebarAppearsTransparent(true);
+
+            if frame == Frame::Buttonless {
+                ns_window.setTitleVisibility(NSWindowTitleHidden);
+            }
```

Comment:
Just FYI, `winit` doesn't guarantee that these aren't overwritten by some other operation done on the window (e.g. when entering or exiting full screen), or that the behaviour here doesn't change subtly in the future (while the previous approach of setting window builder attributes is generally better).

Basically: Use with care.